From: Miguel Vadillo <vadillo@ti.com>
Date: Wed, 13 Apr 2011 22:20:20 +0000 (-0500)
Subject: syslink: ipu_pm: fix kernel warning trace after wd in recovery
X-Git-Tag: android-2.6.35-2.3-omap4.13.1-Beta~44
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=c75157c130ed1c3ae7884a15a0cc0908a6ae8471

syslink: ipu_pm: fix kernel warning trace after wd in recovery

When notifying devh about a wd issue a
blocking_notifier_call_chain was being use,
since the wd is catched in an ISR context a
blocking call is not allowed because it may sleep
causing a kernel warning trace.

Moving the notification and stop_timer to the
already existing clean workqueue is fixing the issue.

Fix error when compiling without WD.

Change-Id: I9a3229cd3c8af31f3b0cf7dec05b219892b44ad6
Signed-off-by: Miguel Vadillo <vadillo@ti.com>

Index: kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c
===================================================================
--- kernel.orig/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-06-19 22:29:19.000000000 +0900
+++ kernel/drivers/dsp/syslink/ipu_pm/ipu_pm.c	2011-06-19 22:31:28.000000000 +0900
@@ -313,6 +313,9 @@
 struct work_struct clean;
 static DECLARE_COMPLETION(ipu_clean_up_comp);
 static bool recover;
+#ifdef CONFIG_DUCATI_WATCH_DOG
+static bool wd_error;
+#endif
 
 /* Latency cstrs */
 #ifdef CONFIG_OMAP_PM
@@ -584,6 +587,20 @@
  */
 static void ipu_pm_clean_work(struct work_struct *work)
 {
+
+#ifdef CONFIG_DUCATI_WATCH_DOG
+	/* Notify devh about WD error */
+	if (wd_error) {
+		ipu_pm_notify_event(0, NULL);
+		wd_error = false;
+	}
+#endif
+
+#ifdef CONFIG_SYSLINK_IPU_SELF_HIBERNATION
+	/* Stop timer */
+	if (sys_rproc->dmtimer != NULL)
+		omap_dm_timer_stop(sys_rproc->dmtimer);
+#endif
 	ipu_pm_clean_res();
 }
 
@@ -3149,10 +3166,8 @@
 		} else if (params->hib_timer_state == PM_HIB_TIMER_WDRESET) {
 			/* notify devh to begin error recovery here */
 			pr_debug("Timer ISR: Trigger WD reset + recovery\n");
+			wd_error = true;
 			ipu_pm_recover_schedule();
-			ipu_pm_notify_event(0, NULL);
-			if (sys_rproc->dmtimer != NULL)
-				omap_dm_timer_stop(sys_rproc->dmtimer);
 			params->hib_timer_state = PM_HIB_TIMER_OFF;
 #endif
 		}
